jax/_src/lru_cache.py

```diff
@@ -0,0 +1,187 @@
+# Copyright 2024 The JAX Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import heapq
+import logging
+import os
+import pathlib
+
+from jax._src import config
+from jax._src.compilation_cache_interface import CacheInterface
+
+
+try:
+  import filelock
+except ImportError:
+  filelock = None
+
+
+logger = logging.getLogger(__name__)
+
+
+class Impl:
```

Comment:
My reason of having this Impl class is that it separates the logic of a LRU cache in general and the compilation-related part. I am thinking that the Impl class can be reused in other places if a LRU cache is needed.

Reply:
I think the only compilation-specific part of LRUCache is reading the max_size config. You could make max_size an argument to `LRUCache.__init__` and combine them if you wanna keep LRUCache general (I agree this is a good idea), or even make max_size optional and default to the config if `None`. I think having the `Impl` is overkill because so little logic is in `LRUCache` right now, and most of it is related to this impl (e.g. checking for supported filesystem)

Reply:
Done